### PR TITLE
Support directly loading gptq models from huggingface

### PR DESCRIPTION
**Pull request by yangw1234**
## Description

Support directly loading gptq models from huggingface.

Many models are published as gptq format in huggingface. It would be nice to load them directly using bigdl-llm.

Install:

```bash
BUILD_CUDA_EXT=0 pip install git+https://github.com/PanQiWei/AutoGPTQ.git@1de9ab6
pip install optimum==0.14.0
```

Usage:
```python
from bigdl.llm.transformers import AutoModelForCausalLM
from transformers import GPTQConfig
quantization_config = GPTQConfig(
    bits=4,
    use_exllama=False,
    )

# Load model in 4 bit,
# which convert the relevant layers in the model into INT4 format
model = AutoModelForCausalLM.from_pretrained(model_path,
                                             load_in_4bit=True, # will load into asym_int4. if using load_in_low_bit, then `load_in_low_bit` must be  "asym_int4"
                                             torch_dtype=torch.float,
                                             trust_remote_code=True,
                                             quantization_config=quantization_config,)
```

Limitations:

Only works on 4bit and asc_order=False
GPU version is really slow. Investigation needed.

Perf: https://github.com/analytics-zoo/nano/issues/738

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/model.py [hunk 1]
```diff
@@ -89,6 +89,18 @@ def from_pretrained(cls,
...
 
         if load_in_4bit or load_in_low_bit:
+
+            if config_dict.get("quantization_config", None) is not None:
+                q_config = config_dict["quantization_config"]
+                if q_config["quant_method"] == "gptq":
+                    invalidInputError(q_config["bits"] == 4,
+                                      "Only 4-bit gptq is supported in bigdl-llm.")
+                    invalidInputError(q_config["desc_act"] is False,
+                                      "Only desc_act=False is supported in bigdl-llm.")
```
Comment (qiuxin2012) [inline python/llm/src/bigdl/llm/transformers/model.py:99]: also check `group_size`, group_size should be a multiple of 64.

File: python/llm/src/bigdl/llm/transformers/model.py [hunk 2]
```diff
@@ -89,6 +89,18 @@ def from_pretrained(cls,
         optimize_model = kwargs.pop("optimize_model", True)
 
         if load_in_4bit or load_in_low_bit:
+
+            if config_dict.get("quantization_config", None) is not None:
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/model.py:93]: do we need to add it to Python Doc?

File: python/llm/example/CPU/HF-Transformers-AutoModels/Model/llama2_gptq/README.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
+# Llama2
```
Comment (jason-dai) [inline python/llm/example/CPU/HF-Transformers-AutoModels/Model/llama2_gptq/README.md:1]: 1) move to `bigdl/python/llm/example/CPU/HF-Transformers-AutoModels/Advanced-Quantizations/GPTQ`
2) `# GPTQ`
3) This example shows how to directly run 4-bit GPTQ models using BigDL-LLM on Intel CPU

File: python/llm/src/bigdl/llm/transformers/convert.py [hunk 4]
```diff
@@ -100,7 +170,25 @@ def _replace_with_low_bit_linear(model, qtype, modules_to_not_convert=None,
...
+                    if is_auto_gptq_available() and isinstance(module, QuantLinearCudaOld):
+                        new_linear = LowBitLinear(
+                            in_features,
+                            out_features,
+                            qtype=qtype,
+                            bias=module.bias is not None,
+                            mp_group=mp_group,
+                        )
+
+                        device_type = module.qweight.data.device.type
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/convert.py:182]: is it used?

File: python/llm/src/bigdl/llm/transformers/model.py (hunk 1 above)
Comment (yangw1234) [reply python/llm/src/bigdl/llm/transformers/model.py:106]: fixed

File: python/llm/src/bigdl/llm/transformers/model.py (hunk 2 above)
Comment (yangw1234) [reply python/llm/src/bigdl/llm/transformers/model.py:99]: Added and passing `quantization_config` is no longer required.

File: python/llm/example/CPU/HF-Transformers-AutoModels/Model/llama2_gptq/README.md (hunk 3 above)
Comment (yangw1234) [reply python/llm/example/CPU/HF-Transformers-AutoModels/Model/llama2_gptq/README.md]: fixed

File: python/llm/src/bigdl/llm/transformers/convert.py (hunk 4 above)
Comment (yangw1234) [reply python/llm/src/bigdl/llm/transformers/convert.py:182]: fixed

File: python/llm/src/bigdl/llm/transformers/model.py [hunk 5]
```diff
@@ -87,8 +92,37 @@ def from_pretrained(cls,
...
+                    if load_in_low_bit is not None:
+                        invalidInputError(load_in_low_bit == "asym_int4",
+                                          "You can only load gptq model as aysm_int4 low bit type.")
+
+                    load_in_low_bit = "asym_int4"
+                    if int(q_config["group_size"]) % get_ggml_qk_size(load_in_low_bit) != 0:
+                        invalidInputError(False,
+                                          (f"group_size must be divisible by "
+                                           f"{get_ggml_qk_size(load_in_low_bit)}."))
+                    if user_quantization_config is not None:
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/model.py:116]: do we want the user to pass `user_quantization_config`?

File: python/llm/src/bigdl/llm/transformers/model.py [hunk 6]
```diff
@@ -87,8 +92,37 @@ def from_pretrained(cls,
...
+                        invalidInputError(user_quantization_config.bits == 4,
+                                          "Only 4-bit gptq is supported in bigdl-llm.")
+                        invalidInputError(user_quantization_config.use_exllama is False,
+                                          "Only use_exllama=False is supported in bigdl-llm.")
+                    else:
+                        from transformers import GPTQConfig
+                        user_quantization_config = GPTQConfig(bits=4, use_exllama=False)
+                    kwargs["quantization_config"] = user_quantization_config
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/model.py:124]: Does save/load low bit work? Do we need to remove `quantization_config` in `save_low_bit`?

File: python/llm/src/bigdl/llm/transformers/model.py (hunk 5 above)
Comment (yangw1234) [reply python/llm/src/bigdl/llm/transformers/model.py:116]: I think not letting user to pass `user_quantization_config` might be a better choice

File: python/llm/src/bigdl/llm/transformers/model.py (hunk 6 above)
Comment (yangw1234) [reply python/llm/src/bigdl/llm/transformers/model.py:125]: save/load low bit works. It seems our `load_low_bit` will ignore `quantization_config`, but I remove it anyway.

**Review by jason-dai — APPROVED** [review]
LGTM

File: python/llm/example/CPU/HF-Transformers-AutoModels/Advanced-Quantizations/GPTQ/README.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+python ./generate.py 
+```
+
+#### 2.2 Server
+For optimal performance on server, it is recommended to set several environment variables (refer to [here](../README.md#best-known-configuration-on-linux) for more information), and run the example with all the physical cores of a single socket.
+
+E.g. on Linux,
+```bash
+# set BigDL-Nano env variables
+source bigdl-nano-init
```
Comment (qiuxin2012) [inline python/llm/example/CPU/HF-Transformers-AutoModels/Advanced-Quantizations/GPTQ/README.md:47]: why nano variables?  we have bigdl-llm-init now.

Comment (yangw1234) [reply python/llm/example/CPU/HF-Transformers-AutoModels/Advanced-Quantizations/GPTQ/README.md:47]: copied from existing examples. will change that.

Comment (yangw1234) [reply python/llm/example/CPU/HF-Transformers-AutoModels/Advanced-Quantizations/GPTQ/README.md:47]: It seems most of our examples still use `bigdl-nano-init`. How about we leave it here and change them together in the future.

Comment (jason-dai) [reply python/llm/example/CPU/HF-Transformers-AutoModels/Advanced-Quantizations/GPTQ/README.md:47]: OK - please open an issue

Comment (yangw1234) [reply python/llm/example/CPU/HF-Transformers-AutoModels/Advanced-Quantizations/GPTQ/README.md:47]: https://github.com/intel-analytics/BigDL/issues/9451

**Comment by yangw1234** [comment]
The failed test is irrelevant and unit tests on arc lack resources to run. I'll merge this PR first to unblock further development.

File: python/llm/example/CPU/HF-Transformers-AutoModels/Advanced-Quantizations/GPTQ/README.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+### 1. Install
+We suggest using conda to manage environment:
+```bash
+conda create -n llm python=3.9
+conda activate llm
+
+pip install bigdl-llm[all] # install bigdl-llm with 'all' option
+pip install transformers==4.34.0
+BUILD_CUDA_EXT=0 pip install git+https://github.com/PanQiWei/AutoGPTQ.git@1de9ab6
+pip install optimum==0.14.0
```
Comment (cyita) [inline python/llm/example/CPU/HF-Transformers-AutoModels/Advanced-Quantizations/GPTQ/README.md:18]: 0.14.0 -》1.14.0

**Pull request merged**